informers.go

```diff
@@ -23,14 +43,122 @@ func getVirtCli() (kubecli.KubevirtClient, error) {
 	return virtCli, err
 }
 
-func getVmiInformer() (cache.SharedIndexInformer, error) {
-	virtCli, err := getVirtCli()
+func getVmiInformer(virtCli kubecli.KubevirtClient) (cache.SharedIndexInformer, error) {
+	listWatcher := cache.NewListWatchFromClient(virtCli.RestClient(), "virtualmachineinstances", namespace, fields.Everything())
+	vmiInformer := cache.NewSharedIndexInformer(listWatcher, &k6tv1.VirtualMachineInstance{}, 1*time.Hour, cache.Indexers{})
+	
+	vmiInformer.AddEventHandler(cache.ResourceEventHandlerFuncs {
+		UpdateFunc: handleVmiUpdate,
+		DeleteFunc: handleVmiDeletion,
+	})
+
+	return vmiInformer, nil
+}
+
+func updateMachineTypes(virtCli kubecli.KubevirtClient) error {
+	vmList, err := virtCli.VirtualMachine(namespace).List(&k8sv1.ListOptions{})
 	if err != nil {
-		return nil, err
+		return err
+	}
+	for _, vm := range vmList.Items {
+		machineType := vm.Spec.Template.Spec.Domain.Machine.Type
+		machineTypeSubstrings := strings.Split(machineType, "-")
+		version := machineTypeSubstrings[2]
+		if len(machineTypeSubstrings) != 3 {
+			return nil
+		}
+		
+		if strings.Contains(version, "rhel") && version < "rhel9.0.0" {
+			machineTypeSubstrings[2] = latestMachineTypeVersion
+			machineType = strings.Join(machineTypeSubstrings, "-")
+			updateMachineType := fmt.Sprintf(`{"spec": {"template": {"spec": {"domain": {"machine": {"type": "%s"}}}}}}`, machineType)
+			
+			_, err = virtCli.VirtualMachine(vm.Namespace).Patch(vm.Name, types.StrategicMergePatchType, []byte(updateMachineType), &k8sv1.PatchOptions{})
+			if err != nil {
+				return err
+			}
+			
+			// add label to running VMs that a restart is required for change to take place
+			if vm.Status.Ready {
+				// adding the warning label to the VMs regardless if we restart them now or if the user does it manually
+				// shouldn't matter, since the deletion of the VMI will remove the label and remove the vmi list anyway
+				err = addWarningLabel(virtCli, &vm)
+				if err != nil {
+					return err
+				}
+				
+				if restartNow {
+					err = virtCli.VirtualMachine(vm.Namespace).Restart(vm.Name, &k6tv1.RestartOptions{})
+					if err != nil {
+						return err
+					}
+				}
+			}
+		}
 	}
+	return nil
+}
 
-	listWatcher := cache.NewListWatchFromClient(virtCli.RestClient(), "virtualmachineinstances", k8sv1.NamespaceAll, fields.Everything())
-	vmiInformer := cache.NewSharedIndexInformer(listWatcher, &k6tv1.VirtualMachineInstance{}, 1*time.Hour, cache.Indexers{})
+func addWarningLabel (virtCli kubecli.KubevirtClient, vm *k6tv1.VirtualMachine) error {
+	addLabel := fmt.Sprint(`{"metadata": {"labels": {"restart-vm-required": "true"}}}}}`)
+	_, err := virtCli.VirtualMachineInstance(vm.Namespace).Patch(vm.Name, types.StrategicMergePatchType, []byte(addLabel), &k8sv1.PatchOptions{})
```

Comment:
IMO we better add the label to the VM object instead of the VMI.
There are two reasons for that:
1. Technically, a VMI object cannot be restarted, only a VM object
2. The user-facing object is the VM object. This is meaningful, since most users are aware (in terms of monitoring, querying data, etc) of the VM object, and the VMI object is treated as an "implementation detail", similarly to the virt-launcher pod.

WDYT?

Reply:
This makes sense to me. I think this might have been what I was originally planning to do (I can't remember why I decided not to but I can't think of a good reason for it).